include/util/cell_storage.hpp

```diff
@@ -0,0 +1,295 @@
+                return *this;
+            }
+
+            bool operator==(const ColumnIterator& other) const
+            {
+                return current == other.current;
+            }
+
+            bool operator!=(const ColumnIterator& other) const
+            {
+                return current != other.current;
+            }
+
+            int operator-(const ColumnIterator &other) const { return (current - other.current) / offset; }
+
+          private:
+            WeightPtrT current;
+            std::size_t offset;
+        };
+
+        std::size_t GetRow(NodeID node) const
+        {
+            return std::find(source_boundary, source_boundary + num_source_nodes, node) -
+                   source_boundary;
+        }
+        std::size_t GetColumn(NodeID node) const
+        {
+            return std::find(
+                       destination_boundary, destination_boundary + num_destination_nodes, node) -
+                   destination_boundary;
+        }
+
+      public:
+        std::pair<RowIterator, RowIterator> GetOutWeight(NodeID node) const
+        {
+            auto row = GetRow(node);
+            auto begin = weights + num_destination_nodes * row;
+            auto end = begin + num_destination_nodes;
+            return std::make_pair(begin, end);
+        }
+
+        std::pair<ColumnIterator, ColumnIterator> GetInWeight(NodeID node) const
+        {
+            auto column = GetColumn(node);
+            auto begin = ColumnIterator{weights + column, num_destination_nodes};
+            auto end =
+                ColumnIterator{weights + column + num_source_nodes * num_destination_nodes,
+                               num_destination_nodes};
+            return std::make_pair(begin, end);
+        }
+
+        auto GetSourceNodes() const
+        {
+            return std::make_pair(source_boundary, source_boundary + num_source_nodes);
+        }
+
+        auto GetDestinationNodes() const
+        {
+            return std::make_pair(destination_boundary,
+                                  destination_boundary + num_destination_nodes);
+        }
+
+        CellImpl(const CellData &data,
+                 WeightPtrT const all_weight,
+                 const NodeID *const all_sources,
+                 const NodeID *const all_destinations)
+            : num_source_nodes{data.num_source_nodes},
+              num_destination_nodes{data.num_destination_nodes},
+              weights{all_weight + data.weight_offset},
+              source_boundary{all_sources + data.source_boundary_offset},
+              destination_boundary{all_destinations + data.destination_boundary_offset}
+        {
+        }
+    };
+
+    inline std::size_t LevelIDToIndex(LevelID level) const { return level-1; }
+
+  public:
+    using Cell = CellImpl<EdgeWeight *>;
+    using ConstCell = CellImpl<const EdgeWeight *>;
+
+    template <typename GraphT>
+    CellStorage(const MultiLevelPartition &partition, const GraphT &base_graph)
```

Comment:
DynamicGraph would work as well, considering we might run this after the graph transformation in the partitioner without creating a new graph. Maybe we should add a graph-trait to solidify the interface?